streams/src/main/java/org/apache/kafka/streams/state/internals/BufferValue.java

```diff
@@ -16,83 +16,162 @@
  */
 package org.apache.kafka.streams.state.internals;
 
+import org.apache.kafka.streams.processor.internals.ProcessorRecordContext;
+
 import java.nio.ByteBuffer;
 import java.util.Arrays;
 import java.util.Objects;
 
 public final class BufferValue {
-    private final ContextualRecord record;
+    private static final int NULL_VALUE_SENTINEL = -1;
+    private static final int OLD_PREV_DUPLICATE_VALUE_SENTINEL = -2;
     private final byte[] priorValue;
+    private final byte[] oldValue;
+    private final byte[] newValue;
```

Comment:
By no longer wrapping and delegating to the ContextualRecord, we are able to store the "old" and "new" parts of the Change independently, which is ultimately what lets us deduplicate the old value and the prior value.